services/src/datasets/external/natur40.rs

```diff
@@ -0,0 +1,290 @@
+    engine::{
+        MetaData, MetaDataProvider, RasterQueryRectangle, RasterResultDescriptor,
+        VectorQueryRectangle, VectorResultDescriptor,
+    },
+    mock::MockDatasetDataSourceLoadingInfo,
+    source::{GdalLoadingInfo, OgrSourceDataset},
+};
+use log::info;
+use reqwest::Client;
+use serde::{Deserialize, Serialize};
+use snafu::ResultExt;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Natur40DataProviderDefinition {
+    id: DatasetProviderId,
+    name: String,
+    base_url: String,
+    user: String,
+    password: String,
+}
+
+#[typetag::serde]
+#[async_trait]
+impl DatasetProviderDefinition for Natur40DataProviderDefinition {
+    async fn initialize(self: Box<Self>) -> crate::error::Result<Box<dyn DatasetProvider>> {
+        Ok(Box::new(Natur40DataProvider {
+            id: self.id,
+            base_url: self.base_url,
+            user: self.user,
+            password: self.password,
+        }))
+    }
+
+    fn type_name(&self) -> String {
+        "Natur4.0".to_owned()
+    }
+
+    fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    fn id(&self) -> DatasetProviderId {
+        self.id
+    }
+}
+
+pub struct Natur40DataProvider {
+    id: DatasetProviderId,
+    base_url: String,
+    user: String,
+    password: String,
+}
+
+#[derive(Deserialize, Debug)]
+struct RasterDb {
+    name: String,
+    title: String,
+    #[serde(deserialize_with = "string_or_string_array")]
+    tags: Vec<String>,
+}
+
+impl RasterDb {
+    fn url_from_name(base_url: &str, name: &str) -> String {
+        format!(
+            "WCS:{}/rasterdb/{name}/wcs?VERSION=1.0.0&COVERAGE={name}",
+            base_url,
+            name = name
+        )
+    }
+
+    fn url(&self, base_url: &str) -> String {
+        Self::url_from_name(base_url, &self.name)
+    }
+}
+
+#[derive(Deserialize, Debug)]
+struct RasterDbs {
+    rasterdbs: Vec<RasterDb>,
+    #[serde(deserialize_with = "string_or_string_array")]
+    tags: Vec<String>,
+    session: String,
+}
+
+#[async_trait]
+impl DatasetProvider for Natur40DataProvider {
+    async fn list(&self, _options: Validated<DatasetListOptions>) -> Result<Vec<DatasetListing>> {
+        // TODO: query the other dbs as well
+        let raster_dbs = self.load_raster_dbs().await?;
+
+        let mut listing = vec![];
+
+        let datasets = raster_dbs
+            .rasterdbs
+            .iter()
+            .map(|db| self.load_dataset(db.url(&self.base_url)));
+        let datasets: Vec<Result<gdal::Dataset>> = join_all(datasets).await;
+
+        for (db, dataset) in raster_dbs.rasterdbs.iter().zip(&datasets) {
+            if let Ok(dataset) = dataset {
+                for band_index in 1..dataset.raster_count() {
+                    if let Ok(result_descriptor) =
+                        raster_descriptor_from_dataset(&dataset, band_index, None)
+                    {
+                        // TODO: get label from rasterband
```

Comment:
Sohaib is trying to get this info?

Reply:
64f671c676b48a0741b2d54adb495a5cbe2abd8e

Reply:
Are there many labels and is this called often?
Since delete(0) is an expensive op for a vector, you can decide if it is worth it changing it to dismiss it in the loop or just leave it as it is.

Reply:
there is one label for each band in the dataset and the datasets have only 1 or 2 bands. For each dataset we perform a WCS request so the delete() should be insignificant